File: src/main/java/com/ioannuwu/inline/ui/render/RustStyleElementRenderer.java
package com.ioannuwu.inline.ui.render;

import com.intellij.openapi.editor.Document;
import com.intellij.openapi.editor.EditorCustomElementRenderer;
import com.intellij.openapi.editor.Inlay;
import com.intellij.openapi.editor.markup.TextAttributes;
import com.ioannuwu.inline.ui.render.elements.graphiccomponents.FontMetricsProvider;
import com.ioannuwu.inline.ui.render.elements.graphiccomponents.FontProvider;
import com.ioannuwu.inline.ui.render.elements.graphiccomponents.GraphicsComponent;
import org.jetbrains.annotations.NotNull;

import java.awt.*;
import java.util.Collection;

public class RustStyleElementRenderer implements EditorCustomElementRenderer {

    private final Collection<GraphicsComponent> graphicComponents;
    private final Document document;

    private final int startOffset;
    private final FontMetricsProvider editorFontMetricsProvider;

    public RustStyleElementRenderer(Collection<GraphicsComponent> graphicComponents, Document document,
                                    int startOffset, FontMetricsProvider editorFontMetricsProvider) {
        this.graphicComponents = graphicComponents;
        this.document = document;
        this.startOffset = startOffset;
        this.editorFontMetricsProvider = editorFontMetricsProvider;
    }

    @Override
    public int calcWidthInPixels(@NotNull Inlay inlay) {
        int sum = 0;
        for (var comp : graphicComponents)
            sum += comp.width();

        return (sum == 0) ? 1 : sum;
    }

    @Override
    public void paint(@NotNull Inlay inlay, @NotNull Graphics g, @NotNull Rectangle targetRegion, @NotNull TextAttributes textAttributes) {
        if (document.getTextLength() < startOffset) return;
        int errorLineNumber = document.getLineNumber(startOffset);
        int lineStartOffset = document.getLineStartOffset(errorLineNumber);

        int offsetFromLineStart = startOffset - lineStartOffset;

        Rectangle newTargetRegion = new Rectangle(targetRegion.x + (offsetFromLineStart) * editorFontMetricsProvider.fontMetrics().charWidth('a'),
                targetRegion.y, targetRegion.width, targetRegion.height);

        for (var comp : graphicComponents) {
            if (comp.width() == 0) comp.draw(inlay, g, newTargetRegion, textAttributes);
        }
        for (var comp : graphicComponents) {
            if (comp.width() != 0) comp.draw(inlay, g, newTargetRegion, textAttributes);
        }
    }
}

File: src/main/kotlin/com/ioannuwu/inline/domain/elements/RustStyleElementRenderer.kt
package com.ioannuwu.inline.domain.elements

import com.intellij.openapi.editor.EditorCustomElementRenderer
import com.intellij.openapi.editor.Inlay
import com.intellij.openapi.editor.markup.TextAttributes
import com.ioannuwu.inline.data.FontData
import com.ioannuwu.inline.domain.NumberOfWhitespaces
import com.ioannuwu.inline.domain.graphics.GraphicsComponentKt
import java.awt.Graphics
import java.awt.Rectangle

/**
 * Indentation level should be calculated from `offsetFromLineStart`
 */
class RustStyleElementRenderer(
    private val graphicsComponents: Collection<GraphicsComponentKt>,
    private val offsetFromLineStart: Int,
    private val editorFontData: FontData,
    private val numberOfWhitespaces: NumberOfWhitespaces,
) : EditorCustomElementRenderer {


    override fun calcWidthInPixels(inlay: Inlay<*>): Int =
        graphicsComponents.sumOf(GraphicsComponentKt::width)

    override fun paint(inlay: Inlay<*>, g: Graphics, targetRegion: Rectangle, textAttributes: TextAttributes) {

        val charWidth = editorFontData.fontMetrics.charWidth('a')

        val newX = targetRegion.x + charWidth * offsetFromLineStart + charWidth * numberOfWhitespaces.numberOfWhitespaces

        val newTargetRegion = Rectangle(newX,targetRegion.y,targetRegion.width, targetRegion.height)

        graphicsComponents.asSequence()
            .sortedBy(GraphicsComponentKt::priority)
            .forEach { it.draw(g, newTargetRegion) }
    }
}